pkg/istio-agent/xds_proxy.go

```diff
@@ -708,3 +721,165 @@ func sendWithTimeout(ctx context.Context, sendFunc func(errorChan chan error)) e
 		return err
 	}
 }
+
+// tapRequest() sends "req" to Istiod, returning a channel through which the requestor gets their reply.
+//
+// Istiod won't reply with a unique ID to let us know when the
+// response has arrived, but will at least return the requested TypeUrl.  We will keep a map of lists of every
+// request for a TypeUrl, and when a response comes in we will alert the requesters first-come, first served.
+func (p *XdsProxy) tapRequest(req *discovery.DiscoveryRequest) (chan *discovery.DiscoveryResponse, error) {
+	if p.connected == nil {
+		return nil, fmt.Errorf("proxy not connected to Istiod")
+	}
+
+	req.ResponseNonce = time.Now().String()
+
+	responseChannel := make(chan *discovery.DiscoveryResponse)
+
+	p.tapMutex.Lock()
+	requestsQueue, ok := p.tapRequests[req.TypeUrl]
+	if !ok {
+		requestsQueue = make([]chan *discovery.DiscoveryResponse, 0, 1)
+	}
+	requestsQueue = append(requestsQueue, responseChannel)
```

Comment:
do we really need multiple concurrently queued requests? or just a single one per-typeurl?

Reply:
actually I am not sure its even possible to send 2 requests in a row without a response? Istiod will block the second request

Reply:
I imagine multiple 3rd party tools pulling at the same time.  I can re-factor the code so that if multiple tools hit the debug tap at once that block on the agent rather than going up together to Istiod.

Reply:
Yeah I think that would work better, and agree its useful to be able to have multiple requests